protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,100 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	"github.com/nats-io/nats.go"
+)
+
+// Protocol is a reference implementation for using the CloudEvents binding
+// integration. Protocol acts as both a NATS client and a NATS handler.
+type Protocol struct {
+	conn *nats.Conn
+
+	Consumer *consumer
+
+	Sender *sender
+
+	connOwned bool // whether this protocol created the nats connection
+}
+
+// NewProtocol creates a new NATS protocol.
+func NewProtocol(ctx context.Context, url, sendSubject string, natsOpts []nats.Option, opts ...ProtocolOption) (*Protocol, error) {
```

Comment:
question: is natsOpts optional (can be nil/0 len) or required? If not, please make it a ProtocolOption otherwise check for its validity and return error early

Reply:
Similarly, is url and sendSubject required or can it be "" ?

Reply:
url is required by both Consumer and Sender.  sendSubject is required by Sender.

Reply:
natsOpts is optional, but is highly used for things like credentials for security.

Reply:
The problem with natsOpts, is that I need it before applying the options in order to get a `*nats.Conn`.  So a sort of chicken-egg problem.  I need the  `*nats.Conn` to create a consumer, but I need a consumer to apply an option.

Reply:
I could set the `*nats.Conn` on the consumer after applying the options, but I would have to see if any other option or logic needs the `*nats.Conn` sooner.

Reply:
If you like, I can add three more options:
`func WithURL(url string, natsOpts []nats.Option) ProtocolOption`
`func WithConnection(conn *nats.Conn) ProtocolOption`
`func WithSenderSubject(sendSubject string) ProtocolOption`

And then I could reduce to only one constructor that takes `(ctx context.Context, opts ...ProtocolOption)`

Of course, there would be more errors needed in case a Sender was created without sendSubject. 
Also "one and only one" of WithURL/WithConnection must be provided.

Reply:
Maybe it's a good choice to keep it as concise as the last one, but it requires comments to clarify the function parameter and validation logic

Reply:
IMHO, since we're starting somewhat from scratch here, the API should be oriented around the use of it: from a sender/receiver perspective. Required arguments should always be part of the signature, anything optional should be a variadic option. Then internally, after all options are applied without throwing individual errors from the option handler, we use a validation function (to keep the constructor clean) to check for more complex validation cases e.g., the "one and only one" case you mention above.

Reply:
I added a validateOptions() method.  There is now only one constructor in Protocol, Receiver, Sender
I also added a WithURL() and WithConnection() option.